Support ADD/DELETE/REMOVE, add eslint, update tests, support different types, update README

petetnt opened the pull request:
Well that sort of escalated quicky.

This PR adds the following features

- ESLinting via eslint-config-motley minus semicolons
- Support for ADD and all kinds of other weird things
- Updated tests and all (4!) tests pass now
- Updated README.md
- Updated package.json
 
The implementation might not be 100% correct nor the
most efficient but hey at least all tests pass now so :shrug:
~~Most likely needs some null/undefined handling too?~~

Signed-off-by: petetnt <pete.a.nykanen@gmail.com>

wolverian commented:
BTW, we can test that this actually works with e.g. https://github.com/trayio/dynamodb-local

petetnt commented:
Rebased against master

wolverian commented on .eslintrc.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  "extends": "motley",
```
dis good

wolverian commented on README.md:
```diff
@@ -19,4 +19,17 @@ const result = await documentClient.update({
+
+## Tests
+
+To run tests with `tape`, run 
```
No need to mention `tape` here I think, it's just an implementation detail.

petetnt replied:
Fixed

wolverian commented on README.md:
````diff
@@ -19,4 +19,17 @@ const result = await documentClient.update({
+
+```
+yarn test # or npm test
+
````
Extra line

petetnt replied:
Fixed

wolverian commented on index.js:
```diff
@@ -1,10 +1,50 @@
+ * @param {string} type - type of item to add to
+ * @returns {{sets, adds}} - Object containing set of adds and sets
+ */
+const merge = (prefix, updates, accumulator, type) => {
```
🙀

petetnt replied:
I know rite

wolverian commented on index.js:
```diff
@@ -1,10 +1,50 @@
+  const addsString = adds.length ? `ADD ${adds.join(", ")}` : ""
+  const setsString = sets.length ? `SET ${sets.join(", ")} ` : ""
+
+  return { UpdateExpression: `${setsString}${addsString}`.trim() }
```
This could be `${setsString} ${addsString}` (note the space) and then we wouldn't need the funny extra space in `setsString` itself. The `trim()` could probably be removed then too?

petetnt replied:
Good call (we need `trim` though so we don't end up with ` ADD foobar biz` when there are no `SET`s

wolverian commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "dynamo-merge",
-  "version": "0.0.3",
+  "version": "1.0.0",
```
I'd rather keep this <1.0.0 until we've tested against DynamoDB.

wolverian commented on package.json:
```diff
@@ -16,7 +16,8 @@
   },
-  "devDependencies": {
+  "dependencies": {
+    "eslint-config-motley": "^4.0.0",
```
Why `dependencies` here?

petetnt replied:
`yarn` (and `airbnb-config-eslint` which `-motley` is based on) default to no external dependencies, that is no devDependencies at all so that all the environments get all the same packages, always.

wolverian replied:
I don't understand what external dependency means here. [The Yarn docs](https://yarnpkg.com/en/docs/managing-dependencies) seem to indicate `yarn add --dev` is supported. [The reference](https://yarnpkg.com/en/docs/package-json#toc-devdependencies) mentions development dependencies too.

wolverian commented on test.js:
```diff
@@ -1,24 +1,33 @@
-        merge({x: 42}), 
-        {UpdateExpression: 'SET x = 42'}
+  t.deepEqual(
+        merge({ x: 42 }),
```
What is this super deep continuation indentation 🙊

wolverian commented:
Thanks for the PR! The functionality is great. I'm not 100% happy with the readability of the code, but I'm leaning towards merging this once my minor nits above are corrected.

petetnt commented:
Fixed the review comment things and added tests and handling for `undefined/null` values and a test for boolean values (do booleans need to be booleans or strings? If booleans then it needs some extra handling in the method).

petetnt commented:
@wolverian I thought the readability sucked too, so I rewrite major parts of it. Should be a bit better now :) 

One could oneline the `getNextPrefix` function expression but it becomes rather messy.

petetnt commented on index.js:
```diff
@@ -1,10 +1,59 @@
+  let newAcc = Object.assign({}, accumulator)
+
+  if (updates === null || typeOfUpdates === "undefined") {
+    return newAcc
```
Maybe we should just return from all the `if´ statements and just skip the `else if`:s
